File: backend/mavito-common-lib/mavito_common/schemas/workspace.py
# mavito-common-lib/mavito_common/schemas/workspace.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List, Dict, Any
from datetime import datetime
import uuid

from mavito_common.models.workspace import GroupType, ItemType


# ========== BookmarkedTerm Schemas ==========


class BookmarkedTermBase(BaseModel):
    """Base schema for bookmarked terms"""
    term_id: uuid.UUID
    notes: Optional[str] = None


class BookmarkedTermCreate(BookmarkedTermBase):
    """Schema for creating a bookmarked term"""
    pass


class BookmarkedTermUpdate(BaseModel):
    """Schema for updating a bookmarked term"""
    notes: Optional[str] = None


class BookmarkedTermResponse(BaseModel):
    """Schema for bookmarked term response"""
    id: uuid.UUID
    user_id: uuid.UUID
    term_id: uuid.UUID
    notes: Optional[str]
    bookmarked_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BookmarkedTermWithDetails(BookmarkedTermResponse):
    """Enriched response with term details for frontend"""
    term: Optional[str] = None
    definition: Optional[str] = None
    domain: Optional[str] = None
    language: Optional[str] = None
    example: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


# ========== BookmarkedGlossary Schemas ==========


class BookmarkedGlossaryBase(BaseModel):
    """Base schema for bookmarked glossaries"""
    domain: str
    notes: Optional[str] = None


class BookmarkedGlossaryCreate(BookmarkedGlossaryBase):
    """Schema for creating a bookmarked glossary"""
    pass


class BookmarkedGlossaryUpdate(BaseModel):
    """Schema for updating a bookmarked glossary"""
    notes: Optional[str] = None


class BookmarkedGlossaryResponse(BookmarkedGlossaryBase):
    """Schema for bookmarked glossary response"""
    id: uuid.UUID
    user_id: uuid.UUID
    bookmarked_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ========== WorkspaceGroup Schemas ==========

class WorkspaceGroupBase(BaseModel):
    """Base schema for workspace groups"""
    name: str
    description: Optional[str] = None
    group_type: GroupType
    color: Optional[str] = None  # Hex color code


class WorkspaceGroupCreate(WorkspaceGroupBase):
    """Schema for creating a workspace group"""
    pass


class WorkspaceGroupUpdate(BaseModel):
    """Schema for updating a workspace group"""
    name: Optional[str] = None
    description: Optional[str] = None
    color: Optional[str] = None


class WorkspaceGroupResponse(WorkspaceGroupBase):
    """Schema for workspace group response"""
    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    items: List["WorkspaceGroupItemResponse"] = []

    model_config = ConfigDict(from_attributes=True)


# ========== WorkspaceGroupItem Schemas ==========

class WorkspaceGroupItemBase(BaseModel):
    """Base schema for workspace group items"""
    item_type: ItemType
    term_id: Optional[uuid.UUID] = None
    domain: Optional[str] = None


class WorkspaceGroupItemCreate(WorkspaceGroupItemBase):
    """Schema for creating a workspace group item"""
    group_id: uuid.UUID


class WorkspaceGroupItemUpdate(BaseModel):
    """Schema for updating a workspace group item"""
    notes: Optional[str] = None


class WorkspaceGroupItemBulkCreate(BaseModel):
    """Schema for bulk creating workspace group items"""
    term_ids: List[uuid.UUID]


class WorkspaceGroupItemResponse(WorkspaceGroupItemBase):
    """Schema for workspace group item response"""
    id: uuid.UUID
    group_id: uuid.UUID
    added_at: datetime
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ========== Search Schemas ==========

class SearchTermsRequest(BaseModel):
    """Schema for searching terms"""
    query: str
    category: Optional[str] = None


class SearchTermsResponse(BaseModel):
    """Schema for search terms response"""
    results: List[Dict[str, Any]]
    total: int

    model_config = ConfigDict(from_attributes=True)


# ========== Bulk Operations ==========

class BulkDeleteRequest(BaseModel):
    """Schema for bulk delete operations"""
    ids: List[uuid.UUID]


# ========== Overview Response ==========

class WorkspaceOverviewResponse(BaseModel):
    """Schema for workspace overview"""
    bookmarked_terms_count: int
    bookmarked_glossaries_count: int
    groups_count: int
    recent_bookmarks: List[BookmarkedTermWithDetails]
    recent_groups: List[WorkspaceGroupResponse]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/workspace.ts
// frontend/src/types/workspace.ts

export interface BookmarkedTerm {
  id: string;
  user_id: string;
  term_id: string;
  notes?: string;
  bookmarked_at: string;
  created_at: string;
  // Term details (included when fetched with term data) - now as direct properties
  term?: string;
  definition?: string;
  domain?: string;
  language?: string;
  example?: string;
}

export interface BookmarkedGlossary {
  id: string;
  user_id: string;
  domain: string;
  notes?: string;
  bookmarked_at: string;
  created_at: string;
}

export interface WorkspaceGroup {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  group_type: GroupType;
  color?: string;
  created_at: string;
  updated_at: string;
  items: WorkspaceGroupItem[];
}

export interface WorkspaceGroupItem {
  id: string;
  group_id: string;
  item_type: ItemType;
  term_id?: string;
  domain?: string;
  added_at: string;
  created_at: string;
}

export const GroupType = {
  TERMS: 'terms',
  GLOSSARIES: 'glossaries',
  MIXED: 'mixed',
} as const;

export const ItemType = {
  TERM: 'term',
  GLOSSARY: 'glossary',
} as const;

export type GroupType = (typeof GroupType)[keyof typeof GroupType];
export type ItemType = (typeof ItemType)[keyof typeof ItemType];

// Request/Response interfaces
export interface BookmarkedTermCreate {
  term_id: string;
  notes?: string;
}

export interface BookmarkedGlossaryCreate {
  domain: string;
  notes?: string;
}

export interface WorkspaceGroupCreate {
  name: string;
  description?: string;
  group_type: GroupType;
  color?: string;
}

export interface WorkspaceGroupUpdate {
  name?: string;
  description?: string;
  color?: string;
}

export interface WorkspaceGroupItemCreate {
  item_type: ItemType;
  term_id?: string;
  domain?: string;
}

export interface WorkspaceGroupItemBulkCreate {
  term_ids: string[];
}

export interface SearchTermsRequest {
  query?: string;
  domain?: string;
  language?: string;
  page?: number;
  limit?: number;
}

export interface SearchTermsResponse {
  terms: Array<{
    id: string;
    term: string;
    definition: string;
    domain: string;
    language: string;
    bookmarked_at: string;
    notes?: string;
  }>;
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface BulkDeleteRequest {
  bookmark_ids?: string[];
  group_ids?: string[];
}

export interface WorkspaceOverview {
  total_bookmarked_terms: number;
  total_bookmarked_glossaries: number;
  total_groups: number;
  recent_terms: Array<{
    id: string;
    term: string;
    definition: string;
    domain: string;
    language: string;
    bookmarked_at: string;
  }>;
}

// UI State interfaces
export interface WorkspaceFilters {
  searchQuery: string;
  selectedDomain: string;
  selectedLanguage: string;
  sortBy: 'date_desc' | 'date_asc' | 'alphabetical';
}

export interface WorkspaceState {
  bookmarkedTerms: BookmarkedTerm[];
  bookmarkedGlossaries: BookmarkedGlossary[];
  groups: WorkspaceGroup[];
  overview: WorkspaceOverview | null;
  filters: WorkspaceFilters;
  selectedItems: string[];
  isLoading: boolean;
  error: string | null;
}
